b1ddi/resource_dns_auth_zone.go

```diff
@@ -416,9 +416,11 @@ func resourceConfigAuthZoneRead(ctx context.Context, d *schema.ResourceData, m i
 	if err != nil {
 		diags = append(diags, diag.FromErr(err)...)
 	}
-	err = d.Set("inheritance_sources", flattenConfigAuthZoneInheritance(resp.Payload.Result.InheritanceSources))
-	if err != nil {
-		diags = append(diags, diag.FromErr(err)...)
+	if resp.Payload.Result.InheritanceSources != nil {
```

Comment:
I'm a bit puzzled about this fix after looking at other objects.

For instance for `resourceConfigViewRead()` we have:
```go
err = d.Set("inheritance_sources", flattenConfigViewInheritance(resp.Payload.Result.InheritanceSources))
if err != nil {
	diags = append(diags, diag.FromErr(err)...)
}

func flattenConfigViewInheritance(r *models.ConfigViewInheritance) []interface{} {
    if r == nil {
        return []interface{}{}
    }
    // ...
}
````

And for `resourceIpamsvcRangeRead()`:
```go
err = d.Set("inheritance_sources", flattenIpamsvcDHCPOptionsInheritance(resp.Payload.Result.InheritanceSources))
if err != nil {
	diags = append(diags, diag.FromErr(err)...)
}

func flattenIpamsvcDHCPOptionsInheritance(r *models.IpamsvcDHCPOptionsInheritance) []interface{} {
    if r == nil {
        return nil
    }
    // ...
}
```

1. Why `dns/auth_zone` would be special and requiring a different handling?
2. Is there any difference when returning `nil` or `[]interface{}{}` in the flattening functions?
3. Is it possible to cover this change with a unit-test?

Reply:
We shouldn't be setting `inheritance_sources` if the API returns an empty object. 
Here is what is happening causing the issue referenced:
- Create an AuthZone with an inheritance set.
- The local terraform state is created with inheritance set in it
- The Read AuthZone API called at the return of the create function doesn't have `inheritance_sources` in its response.
- This causes the local terraform state to update and set the `inheritance_sources` to empty
- If you run `terraform apply` again now, the `inheritance_sources` from the tf file will want to get applied again because of the updates in the previous step. 

To avoid updating `inheritance_sources` erroneously, I added a simple check to update if not nil.


> 2\. Is there any difference when returning `nil` or `[]interface{}{}` in the flattening functions?

Probably no. It should mean the same. I tried setting this to nil and tested it as well. But quickly realized that we shouldn't update `inheritance_sources` if the response is empty or nil.


> 3\. Is it possible to cover this change with a unit-test?

I will try to add one.

Reply:
> We shouldn't be setting `inheritance_sources` if the API returns an empty object. Here is what is happening causing the issue referenced:
> 
>     * Create an AuthZone with an inheritance set.
> 
>     * The local terraform state is created with inheritance set in it
> 
>     * The Read AuthZone API called at the return of the create function doesn't have `inheritance_sources` in its response.
> 
>     * This causes the local terraform state to update and set the `inheritance_sources` to empty
> 
>     * If you run `terraform apply` again now, the `inheritance_sources` from the tf file will want to get applied again because of the updates in the previous step.
> 
> 
> To avoid updating `inheritance_sources` erroneously, I added a simple check to update if not nil.

In this case, isn't the fix to retrieve the `inheritance_sources` when reading the object? If the local state contains it after calling _Create_, how the provider can fix any discrepancy if we don't retrieve the field?

Reply:
https://github.com/infobloxopen/b1ddi-go-client/pull/5 Created this PR in  BloxOne Go SDK repo to support inheritance